pkg/image/controller/trigger/trigger.go

```diff
@@ -0,0 +1,369 @@
+package trigger
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	"k8s.io/kubernetes/pkg/client/record"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/util/metrics"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	imageapi "github.com/openshift/origin/pkg/image/api"
+)
+
+const (
+	// maxRetries is the number of times an image stream will be retried before it is dropped out of the queue.
+	maxRetries = 5
+)
+
+type ImageReaction interface {
+	ImageChanged(obj interface{}, operations []TriggerOperation) error
+}
+
+// TagRetriever returns information about a tag, including whether it exists
+// and the observed resource version of the object at the time the tag was loaded.
+type TagRetriever interface {
+	ImageStreamTag(namespace, name string) (ref string, rv int64, ok bool)
+}
+
+// ImageStreamNamespaceLister helps list and get ImageStreams.
+type ImageStreamNamespaceLister interface {
```

Comment:
This is waiting for Andy's rebase anyway, so generated informers makes this irrelevant.